Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +106,1237 @@ local macro "pderiv_simp" : tactic =>
+  rcases h with ⟨u, rfl⟩
+  exact neg_smul_equiv Q u
+
+lemma neg_of_Z_eq_zero' {P : Fin 3 → R} (hPz : P z = 0) : V.neg P = ![P x, -P y, 0] := by
+  rw [neg, negY_of_Z_eq_zero hPz, hPz]
+
+lemma neg_of_Z_eq_zero {P : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z = 0) : W.neg P =
+    -(Units.mk0 _ (Y_ne_zero_of_Z_eq_zero hP hPz) / Units.mk0 _ (X_ne_zero_of_Z_eq_zero hP hPz))
+      • ![1, 1, 0] := by
+  have hX {n : ℕ} : P x ^ n ≠ 0 := pow_ne_zero n <| X_ne_zero_of_Z_eq_zero hP hPz
+  erw [neg_of_Z_eq_zero' hPz, smul_fin3, Units.val_neg, Units.val_div_eq_div_val, neg_sq, div_pow,
+    (equation_of_Z_eq_zero hPz).mp hP.left, pow_succ, mul_div_cancel_left₀ _ hX, mul_one,
+    Odd.neg_pow <| by decide, div_pow, pow_succ, (equation_of_Z_eq_zero hPz).mp hP.left,
+    mul_div_cancel_left₀ _ hX, Units.val_mk0, mul_one, mul_zero]
+
+lemma neg_of_Z_ne_zero {P : Fin 3 → F} (hPz : P z ≠ 0) : W.neg P =
+    Units.mk0 _ hPz • ![P x / P z ^ 2, W.toAffine.negY (P x / P z ^ 2) (P y / P z ^ 3), 1] := by
+  erw [neg, smul_fin3, mul_div_cancel₀ _ <| pow_ne_zero 2 hPz, ← negY_of_Z_ne_zero hPz,
+    mul_div_cancel₀ _ <| pow_ne_zero 3 hPz, Units.val_mk0, mul_one]
+
+private lemma nonsingular_neg_of_Z_ne_zero {P : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z ≠ 0) :
+    W.Nonsingular ![P x / P z ^ 2, W.toAffine.negY (P x / P z ^ 2) (P y / P z ^ 3), 1] := by
+  exact (nonsingular_some ..).mpr <| Affine.nonsingular_neg <| (nonsingular_of_Z_ne_zero hPz).mp hP
+
+lemma nonsingular_neg {P : Fin 3 → F} (hP : W.Nonsingular P) : W.Nonsingular <| W.neg P := by
+  by_cases hPz : P z = 0
+  · simp only [neg_of_Z_eq_zero hP hPz, nonsingular_smul, nonsingular_zero]
+  · simp only [neg_of_Z_ne_zero hPz, nonsingular_smul, nonsingular_neg_of_Z_ne_zero hP hPz]
+
+variable (V) in
+/-- The negation of a point class. If `P` is a point representative,
+then `V.negMap ⟦P⟧` is definitionally equivalent to `V.neg P`. -/
+def negMap (P : PointClass R) : PointClass R :=
+  P.map V.neg fun _ _ => neg_equiv
+
+lemma negMap_eq {P : Fin 3 → R} : V.negMap ⟦P⟧ = ⟦V.neg P⟧ :=
+  rfl
+
+lemma negMap_of_Z_eq_zero {P : Fin 3 → F} (hP : W.Nonsingular P) (hPz : P z = 0) :
+    W.negMap ⟦P⟧ = ⟦![1, 1, 0]⟧ := by
+  rw [negMap_eq, neg_of_Z_eq_zero hP hPz, smul_eq]
+
+lemma negMap_of_Z_ne_zero {P : Fin 3 → F} (hPz : P z ≠ 0) :
+    W.negMap ⟦P⟧ = ⟦![P x / P z ^ 2, W.toAffine.negY (P x / P z ^ 2) (P y / P z ^ 3), 1]⟧ := by
+  rw [negMap_eq, neg_of_Z_ne_zero hPz, smul_eq]
+
+lemma nonsingularLift_negMap {P : PointClass F} (hP : W.NonsingularLift P) :
+    W.NonsingularLift <| W.negMap P := by
+  rcases P with ⟨_⟩
+  exact nonsingular_neg hP
+
+end Negation
+
+section Addition
+
+/-! ### Addition on point representatives -/
+
+open scoped Classical
+
+variable (V) in
+/-- The addition of two point representatives. -/
+noncomputable def add (P Q : Fin 3 → R) : Fin 3 → R :=
+  if P ≈ Q then ![V.dblX P, V.dblY P, V.dblZ P] else ![V.addX P Q, V.addY P Q, addZ P Q]
```

Comment:
Thanks! I think this and `toAffine_addEquiv` are all I need to complete the reduction to the universal case.